File: crates/arco-cli/src/commands/status.rs
//! Status command - check run status.

use anyhow::{Context, Result};
use clap::Args;
use owo_colors::OwoColorize;

use crate::client::ApiClient;
use crate::{Config, OutputFormat};

/// Arguments for the status command.
#[derive(Debug, Args)]
pub struct StatusArgs {
    /// Run ID to check status for.
    #[arg()]
    pub run_id: Option<String>,

    /// List recent runs instead of checking a specific run.
    #[arg(long, short = 'l')]
    pub list: bool,

    /// Number of runs to list (when using --list).
    #[arg(long, default_value = "10")]
    pub limit: u32,

    /// Show task details.
    #[arg(long, short = 't')]
    pub tasks: bool,
}

/// Execute the status command.
///
/// # Errors
///
/// Returns an error if the workspace ID is missing, arguments are invalid,
/// or the API request fails.
pub async fn execute(args: StatusArgs, config: &Config) -> Result<()> {
    let workspace_id = config
        .workspace_id
        .as_ref()
        .context("Workspace ID is required. Set ARCO_WORKSPACE_ID or use --workspace-id")?;

    let client = ApiClient::new(config)?;

    if args.list {
        list_runs(&client, workspace_id, args.limit, config).await
    } else if let Some(run_id) = &args.run_id {
        show_run(&client, workspace_id, run_id, args.tasks, config).await
    } else {
        anyhow::bail!("Either provide a run ID or use --list to see recent runs")
    }
}

async fn list_runs(
    client: &ApiClient,
    workspace_id: &str,
    limit: u32,
    config: &Config,
) -> Result<()> {
    let response = client.list_runs(workspace_id, Some(limit)).await?;

    match config.format {
        OutputFormat::Json => {
            println!("{}", serde_json::to_string_pretty(&response)?);
        }
        OutputFormat::Text => {
            if response.runs.is_empty() {
                println!("No runs found");
                return Ok(());
            }

            println!("Recent runs:");
            println!();
            for run in &response.runs {
                let state_str = format_state_colored(&run.state.to_string());
                println!(
                    "  {} {} ({}/{} tasks succeeded)",
                    run.run_id, state_str, run.tasks_succeeded, run.task_count
                );
            }
        }
        OutputFormat::Table => {
            use tabled::{Table, Tabled};

            #[derive(Tabled)]
            struct RunRow {
                #[tabled(rename = "Run ID")]
                run_id: String,
                #[tabled(rename = "State")]
                state: String,
                #[tabled(rename = "Tasks")]
                tasks: String,
                #[tabled(rename = "Created")]
                created: String,
            }

            let rows: Vec<_> = response
                .runs
                .iter()
                .map(|r| RunRow {
                    run_id: r.run_id.clone(),
                    state: r.state.to_string(),
                    tasks: format!("{}/{}", r.tasks_succeeded, r.task_count),
                    created: r.created_at.format("%Y-%m-%d %H:%M:%S").to_string(),
                })
                .collect();

            if rows.is_empty() {
                println!("No runs found");
            } else {
                println!("{}", Table::new(rows));
            }
        }
    }

    Ok(())
}

async fn show_run(
    client: &ApiClient,
    workspace_id: &str,
    run_id: &str,
    show_tasks: bool,
    config: &Config,
) -> Result<()> {
    let run = client.get_run(workspace_id, run_id).await?;

    match config.format {
        OutputFormat::Json => {
            println!("{}", serde_json::to_string_pretty(&run)?);
        }
        OutputFormat::Text | OutputFormat::Table => {
            let state_str = format_state_colored(&run.state.to_string());

            println!("Run: {}", run.run_id);
            println!("State: {state_str}");
            println!("Plan ID: {}", run.plan_id);
            println!("Created: {}", run.created_at);
            if let Some(started) = run.started_at {
                println!("Started: {started}");
            }
            if let Some(completed) = run.completed_at {
                println!("Completed: {completed}");
            }
            println!();
            println!("Task Summary:");
            println!("  Total:     {}", run.task_counts.total);
            println!("  Pending:   {}", run.task_counts.pending);
            println!("  Queued:    {}", run.task_counts.queued);
            println!("  Running:   {}", run.task_counts.running);
            println!(
                "  Succeeded: {}",
                format!("{}", run.task_counts.succeeded).green()
            );
            if run.task_counts.failed > 0 {
                println!(
                    "  Failed:    {}",
                    format!("{}", run.task_counts.failed).red()
                );
            } else {
                println!("  Failed:    {}", run.task_counts.failed);
            }
            println!("  Skipped:   {}", run.task_counts.skipped);
            println!("  Cancelled: {}", run.task_counts.cancelled);

            if show_tasks && !run.tasks.is_empty() {
                println!();
                println!("Tasks:");
                for task in &run.tasks {
                    let task_state = format_state_colored(&task.state.to_string());
                    let asset_info = task
                        .asset_key
                        .as_ref()
                        .map_or(String::new(), |k| format!(" ({k})"));
                    println!("  {} {}{}", task.task_key, task_state, asset_info);
                    if let Some(error) = &task.error_message {
                        println!("    Error: {}", error.red());
                    }
                }
            }
        }
    }

    Ok(())
}

fn format_state_colored(state: &str) -> String {
    match state {
        "SUCCEEDED" => state.green().to_string(),
        "FAILED" => state.red().to_string(),
        "RUNNING" => state.blue().to_string(),
        "PENDING" | "QUEUED" => state.yellow().to_string(),
        "CANCELLED" | "TIMED_OUT" | "SKIPPED" => state.dimmed().to_string(),
        _ => state.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_status_args_parsing() {
        use clap::Parser;

        #[derive(Parser)]
        struct TestCli {
            #[command(flatten)]
            args: StatusArgs,
        }

        let cli = TestCli::parse_from(["test", "run-123", "--tasks"]);
        assert_eq!(cli.args.run_id, Some("run-123".to_string()));
        assert!(cli.args.tasks);
        assert!(!cli.args.list);
    }

    #[test]
    fn test_status_args_list() {
        use clap::Parser;

        #[derive(Parser)]
        struct TestCli {
            #[command(flatten)]
            args: StatusArgs,
        }

        let cli = TestCli::parse_from(["test", "--list", "--limit", "20"]);
        assert!(cli.args.list);
        assert_eq!(cli.args.limit, 20);
    }
}

File: python/arco/src/arco_flow/cli/commands/status.py
"""Status command implementation."""

from __future__ import annotations

import time

from rich.console import Console
from rich.live import Live
from rich.table import Table

from arco_flow.cli.config import ArcoFlowConfig, get_config
from arco_flow.client import ApiError, ArcoFlowApiClient

console = Console()
err_console = Console(stderr=True)

_TERMINAL_STATES = {"SUCCEEDED", "FAILED", "CANCELLED", "TIMED_OUT"}


def show_status(
    *,
    run_id: str | None,
    watch: bool,
    limit: int,
) -> None:
    """Execute status command.

    Args:
        run_id: Specific run ID to check.
        watch: Watch status in real-time.
        limit: Number of recent runs to show.
    """
    config = get_config()

    if run_id:
        _show_run_status(config, run_id, watch)
    else:
        _show_recent_runs(config, limit)


def _show_run_status(config: ArcoFlowConfig, run_id: str, watch: bool) -> None:
    """Show status of a specific run.

    Args:
        run_id: The run ID to check.
        watch: Whether to watch for updates.
    """
    console.print(f"[blue]i[/blue] Checking status for run [cyan]{run_id}[/cyan]...")

    try:
        with ArcoFlowApiClient(config) as client:
            run = client.get_run(
                workspace_id=config.workspace_id,
                run_id=run_id,
            ).payload
            if not watch:
                console.print(_render_run_table(run))
                return

            with Live(_render_run_table(run), refresh_per_second=1) as live:
                while True:
                    state = run.get("state")
                    if state in _TERMINAL_STATES:
                        break
                    time.sleep(2)
                    run = client.get_run(
                        workspace_id=config.workspace_id,
                        run_id=run_id,
                    ).payload
                    live.update(_render_run_table(run))
    except ApiError as err:
        err_console.print(f"[red]✗[/red] Status check failed: {err}")
        raise SystemExit(1) from None


def _show_recent_runs(config: ArcoFlowConfig, limit: int) -> None:
    """Show recent runs.

    Args:
        limit: Number of runs to show.
    """
    console.print(f"[blue]i[/blue] Showing {limit} most recent runs...")

    try:
        with ArcoFlowApiClient(config) as client:
            response = client.list_runs(
                workspace_id=config.workspace_id,
                limit=limit,
            ).payload
    except ApiError as err:
        err_console.print(f"[red]✗[/red] Run listing failed: {err}")
        raise SystemExit(1) from None

    runs = response.get("runs", [])
    if not runs:
        console.print("[yellow]![/yellow] No runs found.")
        return

    table = Table(title="Recent Runs")
    table.add_column("Run ID", style="cyan")
    table.add_column("State")
    table.add_column("Created")
    table.add_column("Completed")
    table.add_column("Tasks")
    table.add_column("Failed")

    for run in runs:
        table.add_row(
            run.get("runId", ""),
            run.get("state", ""),
            run.get("createdAt", ""),
            run.get("completedAt", "") or "-",
            str(run.get("taskCount", "")),
            str(run.get("tasksFailed", "")),
        )

    console.print(table)


def _render_run_table(run: dict[str, object]) -> Table:
    run_id = run.get("runId", "")
    state = run.get("state", "")
    title = f"Run {run_id} ({state})"

    parent_run_id = run.get("parentRunId")
    rerun_kind = run.get("rerunKind")
    if parent_run_id:
        title = f"{title} parent={parent_run_id}"
    if rerun_kind:
        title = f"{title} rerun={rerun_kind}"

    table = Table(title=title, show_lines=True)
    table.add_column("Task", style="cyan")
    table.add_column("State")
    table.add_column("Attempt")
    table.add_column("Started")
    table.add_column("Completed")
    table.add_column("Error")

    tasks = run.get("tasks", []) or []
    if not tasks:
        table.add_row("-", "-", "-", "-", "-", "-")
        return table

    for task in tasks:
        table.add_row(
            str(task.get("taskKey", "")),
            str(task.get("state", "")),
            str(task.get("attempt", "")),
            str(task.get("startedAt", "") or "-"),
            str(task.get("completedAt", "") or "-"),
            str(task.get("errorMessage", "") or "-"),
        )

    return table
